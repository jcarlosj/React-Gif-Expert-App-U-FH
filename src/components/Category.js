import React from 'react';

export const Category = ({ category }) => {

    const requestToGiphy = async () => {    //  https://developers.giphy.com/
        const 
            url = `https://api.giphy.com/v1/gifs/search?q=superman&limit=9&api_key=6t2zjzJn3x3LILIFUpF6GKKcR95JTJhK`,
            response = await fetch( url ),
            { data } = await response.json();
        // console.log( data );

        const gifs = data.map( image => {
            return {
                id: image.id,
                title: image.title,
                url: image.images?.downsized_medium.url
            };
        });
        console.log( gifs );
    }
    requestToGiphy();

    return (
        <>
            <h3>{ category }</h3>
        </>
    )
}
