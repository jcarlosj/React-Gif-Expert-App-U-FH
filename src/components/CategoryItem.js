import React from 'react';

export const CategoryItem = ( { title, url } ) => {
    return (
        <div className="card animate__animated animate__bounceIn">
            <img src={ url } alt={ title } title={ title } />        
            <h4>{ title }</h4>
        </div>
   )
}