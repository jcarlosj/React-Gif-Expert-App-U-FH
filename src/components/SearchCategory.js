import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const SearchCategory = ({ setCategories }) => {

    const [ inputValue, setInputValue] = useState( '' );

    const handleInputChange = event => {
        // console.log( event.target.value );
        setInputValue( event.target.value );
    }

    const handleSubmit = event => {
        event.preventDefault();

        if( inputValue.trim().length > 2 ) {
            setCategories( categories => [ ...categories, inputValue ] );
            setInputValue( '' );
        }
        
    }

    return (
        <form
            onSubmit={ handleSubmit }
        >
            {/* <pre>{ JSON.stringify( inputValue, null, 4 ) }</pre> */}
            <input 
                type="text" 
                value={ inputValue }
                onChange={ handleInputChange }
            />
        </form>
    )
}

SearchCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
