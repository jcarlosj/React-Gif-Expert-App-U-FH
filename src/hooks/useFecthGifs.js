import { useState, useEffect } from 'react';

/** Helper */
import { requestToGiphy } from '../helpers/giphyHelper';

/** Custom Hook */
export const useFetchGifs = ( category ) => {

    const [ state, setState ] = useState({
        loading: true,
        data: []
    });

    useEffect( () => {
        requestToGiphy( category )
            .then( gifs => {

                setTimeout( () => {

                    setState({
                        loading: false,
                        data: gifs
                    });
                    console.log( gifs );
                    
                }, 2000 );

            })
            .catch( error => console.log );
    }, [ category ] );

    return state;
}