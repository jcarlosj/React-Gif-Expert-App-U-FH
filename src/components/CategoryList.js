import React from 'react';

/** Hooks */
import { useFetchGifs } from '../hooks/useFecthGifs';

/** Component */
import { CategoryItem } from './CategoryItem';

/** Functional Components */
export const CategoryList = ({ category }) => {

    const { loading, data:images } = useFetchGifs( category );

    return (
        <>
            <h3>{ category }</h3>

            {   loading && <p className="animate__animated animate__flash">Cargando...</p> }

            <div className="card-grid">
            
                {   images.map( image => (
                        <CategoryItem 
                            key={ image.id }
                            { ...image }    //  Spread
                        />
                    ))
                }
                
            </div>
        </>
    )
}
