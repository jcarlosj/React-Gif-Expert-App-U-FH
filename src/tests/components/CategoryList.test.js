import '@testing-library/jest-dom';
import { shallow } from 'enzyme';

import { CategoryList } from '../../components/CategoryList';

import { useFetchGifs } from '../../hooks/useFecthGifs';

jest.mock( '../../hooks/useFecthGifs' );        //  Simula un módulo con una versión simulada automáticamente cuando se requiere

describe( '<CategoryList />', () => {

    const
        category = 'Game of Thrones',
        key = 'uiey82uy82fe';

    test( 'debe desplegar correctamente', () => {

        useFetchGifs.mockReturnValue({         //  Valor que se devolverá siempre que se llame a la función simulada
            loading: true,
            data: []
        });

        const wrapper = shallow(
            <CategoryList 
                key={ key } 
                category={ category } 
            />
        );

        // console.log( wrapper.debug() );

        expect( wrapper ).toMatchSnapshot();    //  Verifica que un valor coincida con la instantánea (Snapshop) más reciente

    });

    test( 'debe mostrar registros cuando las imagenes han cargado \'useFetchGifs\'', () => {

        const gifs = [{
            id: 'iiuyuidfsf454rdf',
            title: 'The Simpsons',
            url: 'http://localhost/cualquier/imagen.jpg'
        }];

        useFetchGifs.mockReturnValue({         //  Valor que se devolverá siempre que se llame a la función simulada
            loading: false,
            data: gifs
        });

        const wrapper = shallow(
            <CategoryList 
                key={ key } 
                category={ category } 
            />
        );

        expect( wrapper ).toMatchSnapshot();    //  Verifica que un valor coincida con la instantánea (Snapshop) más reciente

    }); 

});