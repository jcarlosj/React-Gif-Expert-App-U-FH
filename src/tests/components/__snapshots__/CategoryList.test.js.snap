// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryList /> debe desplegar correctamente 1`] = `
<Fragment>
  <h3>
    Game of Thrones
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Cargando...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`<CategoryList /> debe mostrar registros cuando las imagenes han cargado 'useFetchGifs' 1`] = `
<Fragment>
  <h3>
    Game of Thrones
  </h3>
  <div
    className="card-grid"
  >
    <CategoryItem
      id="iiuyuidfsf454rdf"
      key="iiuyuidfsf454rdf"
      title="The Simpsons"
      url="http://localhost/cualquier/imagen.jpg"
    />
  </div>
</Fragment>
`;
